### PR TITLE
8284637: Improve String.join performance

### PR DESCRIPTION
**Pull request by XenoAmess**
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed

### Issue
 * [JDK-8284637](https://bugs.openjdk.java.net/browse/JDK-8284637): Improve String.join performance

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8169/head:pull/8169` \
`$ git checkout pull/8169`

Update a local copy of the PR: \
`$ git checkout pull/8169` \
`$ git pull https://git.openjdk.java.net/jdk pull/8169/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8169`

View PR using the GUI difftool: \
`$ git pr show -t 8169`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8169.diff">https://git.openjdk.java.net/jdk/pull/8169.diff</a>

</details>

### CODE REVIEW

**Comment by ExE-Boss** [comment]
The issue is now [JDK‑8284637](https://bugs.openjdk.java.net/browse/JDK-8284637 "[JDK‑8284637] Improve String.join performance").

**Comment by XenoAmess** [comment]
> The issue is now [JDK‑8284637](https://bugs.openjdk.java.net/browse/JDK-8284637).

Yes. Just got the time to update it.

**Comment by RogerRiggs** [comment]
Is there jmh test data to quantify the improvement?

**Comment by RogerRiggs** [comment]
Is there jmh test data to quantify the improvement?

File: src/java.base/share/classes/java/lang/String.java [hunk 1]
```diff
@@ -3272,14 +3273,29 @@ static String join(String prefix, String suffix, String delimiter, String[] elem
...
+            el = elements[1];
+            el.getBytes(value, off, coder); off += el.length();
+            final int delimiterLengthWithCoder = (delimiterLength << coder);
+            for (int i = 2; i < size; i++) {
+                System.arraycopy(
+                        value,
+                        delimiterBytesBegin,
+                        value,
+                        off << coder,
+                        delimiterLengthWithCoder
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/lang/String.java:3290]: Is there any benefit to special casing short delimiters?  1..4 bytes?
Arraycopy has a fixed overhead regardless of the delimiter length.
I might expect a big payoff for coder = Latin1 and delimiter length == 1..2  compared to arraycopy.

Comment (RogerRiggs) [inline src/java.base/share/classes/java/lang/String.java:3290]: Is there any benefit to special casing short delimiters?  1..4 bytes?
Arraycopy has a fixed overhead regardless of the delimiter length.
I might expect a big payoff for coder = Latin1 and delimiter length == 1..2  compared to arraycopy.

Comment (XenoAmess) [reply src/java.base/share/classes/java/lang/String.java:3290]: yes I think there should be performance benifitial when delimiter coder be Latin 1 and output not, which would call StringLatin1.inflate for many times, that is a waste.

**Comment by XenoAmess** [comment]
> Is there jmh test data to quantify the improvement?

Not yet as I don't really know how to run a jmh for a package static function...

Should I put the test upon StringJoiner insterad, as it called String.join? 

I would have a try.

**Comment by XenoAmess** [comment]
> Is there jmh test data to quantify the improvement?

@RogerRiggs added jmh now.

Is there any test suite to run it but not run other jmh tests? Or I must invoke the jmh jar using commandline myself to achieve it?

**Comment by cl4es** [comment]
I think you can use or extend the existing `org.openjdk.bench.java.util.StringJoinerBench` JMH for this.

Refer to docs/testing.md for setup help, but you should be able to run specific JMH using the build system with something like `make test TEST=micro:StringJoinerBench.join`

**Comment by XenoAmess** [comment]
JMH results shows not worthy. closed.

**Pull request closed without merging**